code/sonalyze/sonarlog/persistentcluster.go

```diff
@@ -0,0 +1,553 @@
+// date may not exist and will first have to be created.  Then the file may not exist and may also
+// have to be created.
+//
+// There is an invariant that if we know any files in a leaf directory then we know all the files in
+// that directory: we only need to scan the directory once, and if a new file is created it is
+// created via the PersistentCluster, not behind our back.  (It may be possible to be a little
+// resilient about this by using NoCreate flags to file operations and then handling errors, but
+// I've not worried about this.)
+//
+// File data may (eventually) be cached, but that is transparent to the PersistentCluster, it is
+// handled at the level of the LogFile.
+//
+//
+// Memory use.
+//
+// Over time (months) as the server is up, there may be a substantial amount of metadata in memory -
+// the shadow directory trees may become quite large.  For example, most queries will only very
+// rarely request files from more than a few days ago and caching really old stuff may be
+// undesirable.  In principle we would want to purge disused shadow directory tree entries.
+// However, this is hard because we require that there is only ever one LogFile per appendable file,
+// and so we need to prove, before we purge a directory, that no references exist anywhere in the
+// system to any file in the directory (or we risk the directory becoming reinstated and a new
+// LogFile created for one that is already active).  In addition, given the constraints of the date
+// range, we'd only ever be able to purge from the ends of the date range.
+//
+// On the other hand, let's consider how much data we're talking about.  For one year there are 365
+// directories, not an important factor.  But for saga+betzy+fram+fox there are > 2000 nodes, and
+// one file for each per day.  Even if we intern all the strings (and we probably should) then each
+// file structure is at least 64 bytes, and not pointer-free.  2000*365*64=46MB.  I guess it's
+// probably affordable, for now.
+//
+// A halfway solution, should non-purging turn out to be a problem, is to purge *everything* every
+// so often, effectively restarting the daemon.
+
+package sonarlog
+
+import (
+	"fmt"
+	"io/fs"
+	_ "log"
+	"os"
+	"path"
+	"strings"
+	"sync"
+	"time"
+
+	"go-utils/config"
+	"go-utils/hostglob"
+	"go-utils/slices"
+	utime "go-utils/time"
+	. "sonalyze/common"
+)
+
+// Locking discipline (notes).
+//
+// In many contexts, the mutex on the PersistentCluster dominates the mutexes on the files in the
+// cluster.  However, files must be locked individually by all mutable operations.  In a number of
+// cases, file operations may be performed without the cluster lock being held.  The file should not
+// assume that the cluster lock is held (and should ideally not know the cluster it belongs to).
+//
+// Methods that assume the lock is held on self when called are named whateverMethodLocked().
```

Comment:
Document the lock hieararchy either here (or elsewhere and cross-reference from here).

Reply:
going to not do this now.